### ISSUE TITLE
Template resolving discussion.

### ISSUE THREAD

**Issue by danchom** [eclipse/smarthome#3099]
I open this issue to continue discussion from #3050 - about template resolving.
 
@kaikreuzer 's comment:

> I think the PR#3097 will fix the problem. 

Thanks.

> But IMHO this should be a temporary change.

Yes, so we should continue discussing a proper solution. I did a suggestion above already:

> We could also try to dynamically determine this: If a module is marked as HIDDEN, we could keep the placeholder, while any module that is meant to be displayed in the UI should also have "real" config values and no placeholder anymore.

This would allow us to do "black box templates", i.e. create rules from templates, which do not expose their inner workings, but merely offer their configuration to the user. The hidden, internal modules would still have the placeholders, which are only replaced for the RuntimeRule.
On the other hand, it also allows templates as a one-time skeleton for a complex rule, which is created in all its glory and left for further customization to the user - he might change modules, their configuration, etc and thus there must be no placeholders in the module configuration anymore.

What do you think about that?

**Comment by danchom** [comment]
I think separating module into two types complicate rule creation (the user has to explicitly set a module as hidden) and rule execution.
We have internal discussion and our proposal is:
1) add new RuleRegistry method getResolvedRule(String ruleUID) which will return rule with resolved configuration placeholders. In this way you can get rule with placeholders stored into RuleManagedProvider using common RuleRegistry.get() method and resolved rule using the RuleRegistry.getResolvedRule()  method.
2) add new event topic: smarthome/rules/{ruleID}/resolved which will be fired by the ruleRegistry with resolved rule. Again if you want the resolved template rule " in all its glory" you have to register for update topic, otherwise if you just want to show the resolved rule you have to use resolved topic.

**Comment by kaikreuzer** [comment]
Let's move a step back:

Our design of the rule engine never envisioned any placeholder support within rule modules.
The only places where we defined to have placeholders are:
1. rule templates, where placeholders were to be replaced upon rule creation
2. "inner" modules of a composite module - this is just an implementation detail of composite modules; to the outside world these inner modules are not visible and the composite module itself does not have any placeholders in its configuration.

The current discussion about having a configuration on a rule itself (which we didn't have so far either) came up for rule templates that are supposed to keep their inner working hidden and only offer a "gobal" configuration on rule level to the user. In this sense, I see it as a very similar situation to the composite rules: ONLY the hidden parts are allowed to have placeholders, all the rest must not have this.

If you allow placeholders anywhere, you easily get into trouble: See all the issues that were caused by this in the rule editor. And this was not because the rule editor is buggy; no, you simply won't find any sensible UI concepts to deal with the addiitonal complexity of placeholders - at least I wouldn't know how to deal with that in a way that users still understand it. And I do not see how your proposal would in any way address this, sorry.

**Comment by danchom** [comment]
I'm agree with 1 and 2 but I want mention there is a place holders for inputs of the modules too (They are resolved dynamically - on rule execution).
>The current discussion about having a configuration on a rule itself (which we didn't have so far either) came up for rule templates that are supposed to keep their inner working hidden and only offer a "gobal" configuration on rule level to the user. 

Yes that is correct, but how can you configure such rule if you lost the references and replace them with calculated values in configured modules? This is one way rule creation and further configuration is  impossible.  

>If you allow placeholders anywhere, you easily get into trouble: See all the issues that were caused by this in the rule editor. And this was not because the rule editor is buggy; no, you simply won't find any sensible UI concepts to deal with the addiitonal complexity of placeholders - at least I wouldn't know how to deal with that in a way that users still understand it.

For me the API should not be related only to sensible UI concepts, it has to be useful for everyone - to GUI and to advance usage. I agree that paper ui GUI can't  deal with the placeholders, but for that was my proposal. 

> And I do not see how your proposal would in any way address this, sorry.

May be I was not clear. In my proposal the paper ui should use:
1) RuleRegistry.getResolvedRule(ruleUID) instead of RuleRegistry.get(ruleUID)
2) It should be subscribe to smarthome/rules/{ruleID}/resolved topic instead of smarthome/rules/{ruleID}/updated topic

In that way we will keep the placeholders store into  rule managed provider and will have possibility to: 1) to update configuration of rule created by template 
2) to have configuration of rules in feature 
3) to check how the rule is resolved from templates, not to be blackbox and nightmare for debuging 
4) to create a GUI which can work with placeholders 
5) to get the rule in form as it was added. IMHO if you add placeholder you have to get placeholder not something else done by magic.

**Comment by kaikreuzer** [comment]
>  there is a place holders for inputs of the modules too

I wouldn't call this placeholder - this is really rather a reference instead as you actually cannot set a value for an input, but it always only is a reference to some output.

> but how can you configure such rule if you lost the references and replace them with calculated values in configured modules? 

You cannot configure the rule, you can only configure the modules. This was really our fundamental design of templates: They simply create the rule for you initially and then they "disappear".

> For me the API should not be related only to sensible UI concepts, it has to be useful for everyone - to GUI and to advance usage.

Note that all the meta-data has been added only to enable solutions to build GUIs for the rules. So it is not a matter of the Paper UI (which mainly serves as a reference/demo implementation and for testing), but if you introduce a burden/showstopper for UIs that would apply for ANY UI that are built on top of this architecture.
And yes, my suggestion was exactly to enable both (UI and "advanced" usage): Note that the "visibility" flag has been exclusively added for UI support. So if we mark the parts of a rule as "hidden", which are simply too "advanced"/complex to be supported in a meaningful way in a UI, then everyone is served, isn't it?

> the paper ui and everyone who needs of resolved plaseholders should use RuleRegistry.getResolvedRule(ruleUID)

Try it out yourself in your head: It simply won't work that way as the Paper UI will naturally edit the value and thus, once the edited rule is saved, all your placeholders will be gone, so you didn't win anything.

**Comment by danchom** [comment]
>I wouldn't call this placeholder - this is really rather a reference instead as you actually cannot set a value for an input, but it always only is a reference to some output.

For me the placeholders are equivalent to references and we have applied same concept  for module configuration values as the input values.

>You cannot configure the rule, you can only configure the modules. This was really our fundamental design of templates: They simply create the rule for you initially and then they "disappear".

Yes my question was how to configure modules if we have lost references in them.
Our initial design was:
template(paramA), paramA = 5 -> ruleA(){condition(x < 5)} in rule registry -> ruleA(){condition(x < 5)} in rule engine

To change the paramA to be 6 for the rule created in that way you have two options:
1) remove the rule, create a new rule instance from template applying to it a new property value =6 and then add the rule.
2) get the created rule, find the condition, update configuration of condition and update the rule. If you do that in this way you must have knowledge about the rule i.e the template is not a black box 

In the last big refactoring we have change that :
template(paramA), paramA = 5 -> ruleA(paramA){condition(x < ${paramA})} in rule registry -> ruleA(){condition(x < 5)} in rule engine

 In that way if you want to update template's configuration value, which becomes same as to update the rule's configuration value, you have to do:
get ruleA, update the configuration of rule (by calling ruleA.getConfiguration().put(paramA, 6)) and then update the rule. 

>Note that all the meta-data has been added only to enable solutions to build GUIs for the rules.

For me the GUI is just a corner case of meta-data usage. If the meta-data is applicable to automation module and other applications, why not to use it?

>but if you introduce a burden/showstopper for UIs that would apply for ANY UI that are built on top of this architecture.

I don't want to burden the GUI. IMHO changing the method names is not a big deal. In that way you have clear concept: if you want to work the added rule (the rule "created in all its glory and left for further customization to the user" as you said) you have to use rule registry get method. If you want to get only the result of actual resolved rule executed by the rule engine you have to use the additional getResolvedRule() method. Even more in feature you can create a GUI which can permit update configuration values of rules.
Actually I don't like the idea  to duplicate update event with the resolved one. May be the update event has to contain always resolved values. What do you think?

>Try it out yourself in your head: It simply won't work that way as the Paper UI will naturally edit the value and thus, once the edited rule is saved, all your placeholders will be gone, so you didn't win anything.

Agree, at the moment the PaperUI overwrites existing reference. But this is just a GUI implementation and IMHO this implementation should not dictate the automation design.

**Comment by danchom** [comment]
>if you introduce a burden/showstopper for UIs that would apply for ANY UI that are built on top of this architecture.

Actually If you think this is a breakable change for PaperUI , we can reverse the methods i.e.
get() - will return the resolved rule 
getRule() will return rule with references.
but IMHO this will be confused for the users because they put a rule in ruleregistry and gets something different - resolved rule.
BTW I don't understand your approach with HIDDEN markers. In my example above if we mark a condition as hidden, it will not be appeared in gui and the user will be confused why the rule is not executed or the entire rule should not appear in gui. Could you explain the approach in more details?

**Comment by kaikreuzer** [comment]
> To change the paramA to be 6 for the rule

Note that our design did not at all have the possibility of config params on a rule. So this question is simply not to be asked. The template instead is used to create a rule "based on" (and not "with") certain config values. So the configuration belongs to the template, not to the rule!

> Actually If you think this is a breakable change for PaperUI , we can reverse the methods 

Yes, I would actually suggest to revert the changes that had been done (treating config params as params of the rule and not of the template all of a sudden), so that we have the solution that worked smoothly over the past months. Could you create a PR for that? That would give us some time to discuss and implement the "configs for rules" - note that I will need working templates in the UI next week as I prepare two demos for Eclipse IoT day and Devoxx US...

> In my example above if we mark a condition as hidden, it will not be appeared in gui and the user will be confused why the rule is not executed or the entire rule should not appear in gui.

Why should the user be confused? That was exactly the purpose of visibility=HIDDEN, wasn't it? Or what was your assumption about the effect of that flag?

**Comment by danchom** [comment]
>Note that our design did not at all have the possibility of config params on a rule. 

Yes, but we have talk about rule's configuration which has to be implemented in feature and this is just an extension of our design. IMHO it is very useful feature which we are using. We can be backup compatible with my second proposal: to reverse the method methods. The reference will be stored in rule registry, but the get method will resolve them on each call. We will have second method for example getRule() which will return the rule without resolve references (as they are stored in RuleRegistry).

> Could you create a PR for that? 
Actually my PR  #3050 do this for the templates. The references are resolved when a rule by template is added and the resolved values are stored in RR.

>Why should the user be confused? 

Because he will see only trigger and action and will wonder why the rule is not executed when the trigger is triggered.  He won't know that the condition was not satisfied because there is no condition at all.

**Comment by kaikreuzer** [comment]
> We can be backward compatible with my second proposal: to reverse the methods.

I don't see how that would work, because if you use a UI that sets a configuration value, you will ALWAYS lose your placeholder in there.

> Actually my PR #3050 do this for the templates. The references are resolved when a rule by template is added and the resolved values are stored in RR.

Sorry, yes, you are right. The only thing that remains there is the configuration on the rule, but as long as #3104 is not merged, that's not an issue.

> He won't know that the condition was not satisfied because don't see a condition at all.

Yes, correct. When things are hidden from sight, they cannot be seen :-)

But as I asked above already: What was your assumption about the effect of that flag? Does it at all make sense then? I might also be ok to say that we can only hide all modules at once, which would then allow to have "black box" rules, which only show a configuration to the user that he can edit, but nothing else. Would that be ok?

**Comment by danchom** [comment]
>I don't see how that would work, because if you use a UI that sets a configuration value, you will ALWAYS lose your placeholder in there.

There is two options:
1) Change the GUI in a way that it has to resolve the references and show the values instead of placeholder, but it must not change the rule. When update a module's configuration value which has a reference to a rule's configuration property,  the new value has to be put into rule's configuration (where the reference is pointed) instead of module's configuration. The other option of this approach is the module's configuration properties which referred to rule's configuration to be read only and their values can be changed only by the rule's configuration fields. 

2) Leave the GUI as it is, but when a rule is updated, the RuleRegistry must compare all modules between the updated rule and the stored one and for each module which has a reference to rule's configuration property in the stored rule, it must restore the reference and update the rule's configuration in the updated rule.. 

For me the better approach is 1.

>which would then allow to have "black box" rules, which only show a configuration to the user that he can edit, but nothing else. Would that be ok?

Using visibility in that way can be useful for me. But I think this can be control from rule's visibility not from the module's visibility. For me the module's visibility can be useful only for hiding the modules containing in composite module types.

**Comment by kaikreuzer** [comment]
> But I think this can be control from rule's visibility not from the module's visibility.

Not really, because if you set a rule to HIDDEN, I would expect that you cannot see the rule at all. This is something different then hiding the modules of a rule. So I think it is still ok to use the visibility of modules (and we can say that it usually makes sense to have all or none visible).

**Comment by danchom** [comment]
OK lets leave the visibility property in modules, but how we proceed with the references in the rule. What do you think about my proposal?  I prefer 1, but the 2 will be backward compatible.

**Comment by danchom** [comment]
We have internal discussion and have a new proposal for rule's configuration:
1) When add, update a rule, configuration references will be stored in the managed provider, but they always will be resolved in registry. The resolution will be done by the onAddElement/onUpdateElement which are called after the storing the rule. In that way the user/gui will always work with the resolved rules.
2) We can introduce a new method (now or on in feature) in the ruleRegistry i.e. updateConfiguration(String ruleUID, Map newConfiguration) which will have opportunity to retrieve the rule from managedProvider and try to resolve it again with new configuration values.
I think this is a backward compatible approach which will have additional functionality to update rule's by updating their configuration values.  
Of course when the user/gui try to update a rule with references, passing a resolved rule, it will overwrite the existing references, but the update command works in that way. This approach just gives opportunity to update a rule with references in another way: through the updateConfiguration method. 
@kaikreuzer  what do you think?

**Comment by kaikreuzer** [comment]
Sorry, due to the Easter break I lost this from my view.

I am not fully clear on the implications of your suggestion. It actually means that if you do a `ruleRegistry.update(ruleRegistry.get(rule))` you will end up with a different system then before, am I correct? If so, this sounds highly unexpected and dangerous to me.

**Comment by danchom** [comment]
>  It actually means that if you do a ruleRegistry.update(ruleRegistry.get(rule)) you will end up with a different system then before

Actually not, it will replace in the storage a rule stored with references with the same rule with resolved references (as it works at the moment). From user point of view it is the same rule and same system, but the rule can't be configure anymore because the references are gone. Btw I think this is not a problem because the update command do this. It updates one rule (with config references) with another rule (without references)
If you want to reconfigure a rule with config references, you have to use a new updateConfiguration(String ruleUID, Map newConfiguration) method instead of update method. - it is not implemented yet. This method will get from ManagedProvider a rule with references and will update and store only rule's configuration values. After that it will pass the rule to rule registry to resolve the references again with the new configuration values. In that way after restart the rule will be the same as last configuration update.
The PR #3307 is not implemented this functionality it just fix resolution of config properties:
- now not only rule template are resolved, but a rule with configuration also are resolved.
- fixes applying default value of configuration property defined in template. 
- it improve normalization to be done only one time.

**Comment by kaikreuzer** [comment]
Sorry, but I am completely lost now.
1+2 do not make any sense to me - we talked about having configurations on rule level as a new feature, but 1+2 would destroy any potential use of them; if every placeholder is resolved anyhow, we would not have to store the configuration in the first place.

> From user point of view it is the same rule and same system, but the rule can't be configure anymore because the references are gone

Not it is not - it is a different rule now, which only behaves (currently) in the same way as the previous rule. And that's my point: `ruleRegistry.update(ruleRegistry.get(rule))`should really always give you the very same rule back (with the exception of maybe updated lastmodified timestamps) - everything else is a very unexpected side effect, which I think is a very bad smell.

The registry should expose the entities exactly in the way how the providers deliver them - this is what the `Registry´ interface is about.

**Comment by kaikreuzer** [comment]
As a solution, I would still stick with the initial one:

> If a module is marked as HIDDEN, we could keep the placeholder, while any module that is meant to be displayed in the UI should also have "real" config values and no placeholder anymore.

If you don't like that approach as it is too implicit, we might make the "to-be-resolved" parameters explicit within the rule templates - e.g. by introducing two different config descriptions to them instead of a single one as right now.

**Comment by danchom** [comment]
>The registry should expose the entities exactly in the way how the providers deliver them - this is what the `Registry´ interface is about.

I agree with you, and for this reason we have returned a rule containing references to rule's configuration properties (which is not display well in the GUI). We have treated rule template and rule with configuration in same way. May be the aproach can be to resolve the configuration properties related to templates and leave the rest configuration properties with references. I think this is your last propolsal. We can use i.e. "advanced"=true in config description parameter to mark a property as unresolvable. 
I don't like "HIDDEN" aproach beacause the user can't see all the modules and can't understand the behaviour of the rule. For example if we hide a condition which is not satisfied the user will wonder why the rule is not executed.

**Comment by danchom** [comment]
It is obvious that with the current api there is no clean approach for rule configuration topic. After internal discussion we have agree that:
1) this command invocation ruleRegistry.update(ruleRegistry.get(rule)) must not change the rule
2) it is not good idea a property value to have possibility to contain different value types (i.e. Integer value or String reference to Integer value)
3) When we want to extract module properties as rule configuration property, for easy rule configuration, can create a template with this properties. Actually the templates are for this. 
For these reasons we refuse of rule configuration and I think this issue and PR #3307 can be closed. 
@kaikreuzer what do you think?

**Comment by kaikreuzer** [comment]
Sorry @danchom I have only returned today from vacation.
I'd like to reopen the issue as I do not see any solution for my requirement that I described in https://github.com/eclipse/smarthome/issues/3050#issuecomment-283391284 as "black box templates" (unless I have missed anything in your answers?).

> I don't like "HIDDEN" aproach beacause the user can't see all the modules and can't understand the behaviour of the rule. For example if we hide a condition which is not satisfied the user will wonder why the rule is not executed.

You should note that the HIDDEN feature on modules is in place since a long time and it itself hasn't been part of this discussion (imho). So having a rule that hides a condition is possible in any case already - this is nothing that would be introduced only by my suggestion.

**Comment by danchom** [comment]
I have a question to the "black box templates" requirement. Does the black box rules have  opportunity to be updated with change configuration values or the configuration will be used only to display the values used for rule creation. 
In first case we must save the references between the modules and rule configuration. They can be  reused when the configuration is updated. This leads to following problems: property's get()  method of property of decimal type can returned different type values base on the situation if it contains value (BigDecimal) or String reference. IMHO this check must not be leaved to the user. In this case may be we have to extend Configuration class and add method getRef(key) method which will return a String ref value if it exists or null in other cases. The get() method should always return a resolved value.
In the second case we can leave the rule configuration and show it to the user. The update operation can be done only by template. Remove the existing rule and create a new one (with same UID) by the specified template. I think this is clear and simple approach, otherwise the rule with references will duplicate the logic of templates.
